driver/src/orderbook/streamed/state.rs

```diff
@@ -0,0 +1,531 @@
+            Event::TokenListing(event) => self.token_listing(event),
+            Event::OrderPlacement(event) => self.order_placement(event),
+            Event::OrderCancellation(event) => self.order_cancellation(event, block_batch_id),
+            Event::OrderDeletion(event) => self.order_deletion(event),
+            Event::Trade(_event) => unimplemented!(),
+            Event::TradeReversion(_event) => unimplemented!(),
+            Event::SolutionSubmission(_event) => unimplemented!(),
+        }
+    }
+
+    fn deposit(&mut self, event: &Deposit, block_batch_id: BatchId) -> Result<(), Error> {
+        let balance = self.balances.entry((event.user, event.token)).or_default();
+        balance.deposit(
+            Flux {
+                amount: event.amount,
+                batch_id: event.batch_id,
+            },
+            block_batch_id,
+        );
+        Ok(())
+    }
+
+    fn withdraw_request(&mut self, event: &WithdrawRequest) -> Result<(), Error> {
+        let balance = self.balances.entry((event.user, event.token)).or_default();
+        balance.pending_withdraw = Some(Flux {
+            amount: event.amount,
+            batch_id: event.batch_id,
+        });
+        Ok(())
+    }
+
+    fn withdraw(&mut self, event: &Withdraw, block_batch_id: BatchId) -> Result<(), Error> {
+        let balance = self.balances.entry((event.user, event.token)).or_default();
+        balance.withdraw(event.amount, block_batch_id)
+    }
+
+    fn token_listing(&mut self, event: &TokenListing) -> Result<(), Error> {
+        self.tokens.0.push((event.id, event.token));
+        Ok(())
+    }
+
+    fn order_placement(&mut self, event: &OrderPlacement) -> Result<(), Error> {
+        let order = Order {
+            buy_token: event.buy_token,
+            sell_token: event.sell_token,
+            valid_from: event.valid_from,
+            valid_until: event.valid_until,
+            price_numerator: event.price_numerator,
+            price_denominator: event.price_denominator,
+            used_amount: 0,
+        };
+        match self.orders.entry((event.owner, event.index)) {
+            Entry::Vacant(entry) => entry.insert(order),
+            Entry::Occupied(_) => return Err(Error::OrderAlreadyExists),
+        };
+        Ok(())
+    }
+
+    fn order_cancellation(
+        &mut self,
+        event: &OrderCancellation,
+        block_batch_id: BatchId,
+    ) -> Result<(), Error> {
+        let order = self
+            .orders
+            .get_mut(&(event.owner, event.id))
+            .ok_or(Error::UnknownOrder)?;
+        order.valid_until = block_batch_id - 1;
+        Ok(())
+    }
+
+    fn order_deletion(&mut self, event: &OrderDeletion) -> Result<(), Error> {
+        // Orders are allowed to be deleted multiple times.
+        self.orders.remove(&(event.owner, event.id));
+        Ok(())
+    }
+}
+
+#[derive(Clone, Copy, Debug, Error)]
+pub enum Error {
+    #[error("unknown token")]
+    UnknownToken,
```

Comment:
For debugging purposes it might be nice if these had the IDs as part of the enum type and particularly the error message.

E.g.     `#[error("Unknown Token `{0}`")]`

Reply:
This is implemented in the next PR.